master/internal/db/migrations.go

```diff
@@ -1,37 +1,196 @@
+			if err != nil {
+				return nil, errors.Wrapf(err, "unable to read CA file %q", caPath)
+			}
+
+			if !caCertPool.AppendCertsFromPEM(caCert) {
+				return nil, errors.Wrap(err, "unable to add CA to cert pool")
+			}
+
+			opts.TLSConfig.RootCAs = caCertPool
+			opts.TLSConfig.ClientCAs = caCertPool
+		}
+	}
+
+	return opts, nil
+}
+
+func tablesExist(tx *pg.Tx, tableNames []string) (map[string]bool, error) {
+	existingTables := []string{}
+	result := map[string]bool{}
+	for _, tn := range tableNames {
+		result[tn] = false
 	}
-	m, err := migrate.NewWithDatabaseInstance(migrationURL, "postgres", driver)
+
+	_, err := tx.Query(
+		&existingTables,
+		`SELECT tablename FROM pg_tables WHERE schemaname = 'public' AND tablename IN (?)`,
+		pg.In(tableNames),
+	)
 	if err != nil {
-		return errors.Wrapf(err, "error constructing Postgres migration using %s", migrationURL)
+		return result, err
+	}
+
+	for _, tn := range existingTables {
+		result[tn] = true
 	}
 
-	migrateVersion, _, merr := m.Version()
-	if merr != nil {
-		if merr != migrate.ErrNilVersion {
-			return errors.Wrap(merr, "error loading golang-migrate version")
+	return result, nil
+}
+
+func ensureMigrationUpgrade(tx *pg.Tx) error {
+	exist, err := tablesExist(tx, []string{"gopg_migrations", "schema_migrations"})
+	if err != nil {
+		return err
+	}
+
+	// On fresh installations, just run init.
+	if !exist["gopg_migrations"] && !exist["schema_migrations"] {
+		_, _, err = migrations.Run(tx, "init")
+		return err
+	}
+
+	if exist["gopg_migrations"] || !exist["schema_migrations"] {
+		return nil
+	}
+
+	log.Infof("upgrading migration metadata...")
+
+	type GoMigrateEntry struct {
+		Version string
+		Dirty   bool
+	}
+
+	rows := []GoMigrateEntry{}
+	if _, err = tx.Query(&rows, `SELECT version, dirty FROM schema_migrations`); err != nil {
+		return err
+	}
+
+	// Unrecognized table state.
+	if len(rows) != 1 {
+		return errors.New(fmt.Sprintf("schema_migrations table has %d entries", len(rows)))
+	}
+
+	goMigrateEntry := rows[0]
+
+	if goMigrateEntry.Dirty {
+		return errors.New(
+			fmt.Sprintf("schema_migrations entry dirty, version %s", goMigrateEntry.Version))
+	}
+	goMigrateVersion, err := strconv.ParseInt(goMigrateEntry.Version, 10, 64)
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("Upgrade debug %s %t\n", goMigrateEntry.Version, goMigrateEntry.Dirty)
+
+	// CREATE gopg_migrations table,
+	// and INSERT the initial version from go-migrate.
+	if _, _, err := migrations.Run(tx, "init"); err != nil {
+		return err
+	}
+	if err := migrations.SetVersion(tx, goMigrateVersion); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Migrate runs the migrations from the specified directory URL.
+func (db *PgDB) Migrate(migrationURL string, actions []string) error {
+	// go-pg/migrations uses go-pg/pg connection API, which is not compatible
+	// with pgx, so we use a one-off go-pg/pg connection.
+	pgOpts, err := makeGoPgOpts(db.url)
```

Comment:
seems this could be a method on `db *PgDB` and we could store more than just `url` on it to avoid all the regex'ing, etc. but not a huge deal.

Reply:
I think *ideally* we either patch ParseURL in the upstream to make this not as bad, or make our own migrations lib where this is not necessary at all. Then having a specialized `PgOpts` construction in `*PgDB` would have to be taken out... And so I went for this more compartmentalized approach with all migrations code (and hacks) limited to this one file.